chainer/testing/function.py

```diff
@@ -0,0 +1,530 @@
+
+        Args:
+            outputs_template (tuple of arrays):
+                A set of template arrays.
+                The shapes and dtypes of the returned arrays should match
+                these arrays.
+
+        Returns: A tuple of grad_outputs arrays.
+        """
+        grad_outputs = tuple(
+            numpy.random.uniform(-1, 1, _.shape).astype(_.dtype)
+            for _ in outputs_template)
+        return grad_outputs
+
+    @_copy_to_test_case()
+    def generate_grad_grad_inputs(self, inputs_template):
+        """Generates a set of grad_grad_inputs to test the function.
+
+        The default implementation returns uniform [-1, 1] random values.
+
+        Args:
+            inputs_template (tuple of arrays):
+                A set of template arrays.
+                The shapes and dtypes of the returned arrays should match
+                these arrays.
+
+        Returns: A tuple of grad_grad_inputs arrays.
+        """
+        grad_grad_inputs = tuple(
+            numpy.random.uniform(-1, 1, _.shape).astype(_.dtype)
+            for _ in inputs_template)
+        return grad_grad_inputs
+
+    @_copy_to_test_case()
+    def _to_noncontiguous_as_needed(self, contig_arrays):
+        if self.contiguous is None:
+            # non-contiguous
+            return array_module._as_noncontiguous_array(contig_arrays)
+        if self.contiguous == 'C':
+            # C-contiguous
+            return contig_arrays
+        assert False, (
+            'Invalid value of `contiguous`: {}'.format(self.contiguous))
+
+    @_copy_to_test_case()
+    def _generate_inputs(self):
+        inputs = self.generate_inputs()
+        _check_array_types(inputs, _cpu.CpuDevice(), 'generate_inputs')
+        return inputs
+
+    @_copy_to_test_case()
+    def _generate_grad_outputs(self, outputs_template):
+        grad_outputs = self.generate_grad_outputs(outputs_template)
+        _check_array_types(
+            grad_outputs, _cpu.CpuDevice(), 'generate_grad_outputs')
+        return grad_outputs
+
+    @_copy_to_test_case()
+    def _generate_grad_grad_inputs(self, inputs_template):
+        grad_grad_inputs = self.generate_grad_grad_inputs(inputs_template)
+        _check_array_types(
+            grad_grad_inputs, _cpu.CpuDevice(), 'generate_grad_grad_inputs')
+        return grad_grad_inputs
+
+    @_copy_to_test_case()
+    def _forward_expected(self, inputs):
+        outputs = self.forward_expected(inputs)
+        _check_array_types(outputs, _cpu.CpuDevice(), 'forward_expected')
+        return outputs
+
+    @_copy_to_test_case()
+    def _forward(self, inputs, backend_config):
+        assert all(isinstance(a, chainer.Variable) for a in inputs)
+        with backend_config:
+            outputs = self.forward(inputs, backend_config.device)
+        _check_variable_types(outputs, backend_config.device, 'forward')
+        if not isinstance(outputs, tuple):
+            outputs = outputs,
```

Comment:
This code is never executed since `_check_variable_types` throws in case `outputs` is not a `tuple`. Should we simply remove the above two lines?

Reply:
Fixed.